File: src/leetcode/java/FindMinimumInRotatedSortedArray.java
package leetcode.java;

public class FindMinimumInRotatedSortedArray {
    /**
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public int findMin(int[] nums) {
        int answer = 0;
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            if (nums[left] < nums[right]) {
                answer = Math.min(answer, nums[left]);
                break;
            }

            int mid = (left + right) / 2;
            answer = Math.min(answer, nums[mid]);
            if (nums[mid] >= nums[left]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return answer;
    }
}

File: src/leetcode/kotlin/FindMinimumInRotatedSortedArray.kt
package leetcode.kotlin

class FindMinimumInRotatedSortedArray {
    fun solution(nums: IntArray): Int {
        if (nums.size == 1) {
            return nums[0]
        }

        var left = 0
        var right = nums.size - 1

        if (nums[left] < nums[right]) {
            return nums[0]
        }

        while (right >= left) {
            var mid = left + (right - left) / 2

            if (nums[mid] > nums[mid + 1]) {
                return nums[mid + 1]
            }

            if (nums[mid - 1] > nums[mid]) {
                return nums[mid]
            }

            if (nums[mid] > nums[0]) {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }

        return Int.MAX_VALUE
    }
}